File: modules/prompts.py
"""
Prompts for the Dream Analysis Application.
This module contains template prompts for the LLM at different stages of the pipeline.
"""

# Prompt for generating the dream analysis
DREAM_ANALYSIS_PROMPT = """
As a dream interpreter, analyze the following dream information:

Dream narrative: {narrative}
Main symbols or elements: {symbols}
Primary emotion felt: {emotion}
Emotional intensity (1-5): {intensity}
Connection to waking life: {life_connection}

Provide a structured analysis with the following components:

1. Interpretation: Give a thoughtful overall interpretation of what this dream might mean, focusing on emotional insights.

2. Symbols: Analyze each of these key symbols from the dream:
{symbol_list}
For each symbol, provide:
- A title for the symbol
- An explanation of what the symbol might represent in the dreamer's psyche

3. Advice: Offer actionable guidance based on the dream's emotional underpinnings. What steps might the dreamer take to address the underlying feelings or situations?

4. Image Prompt: Create a vivid visual description that captures the essence of this dream using colors, shapes, and imagery. The visual description should incorporate the emotional tone of the dream.

Consider the personal context shared by the dreamer and avoid generic interpretations. Be insightful yet concise.
"""

# Prompt for generating the image from the dream
IMAGE_GENERATION_PROMPT = """
Create a dreamlike visualization of the following dream:

{image_prompt}

The dream evoked feelings of {emotion} with an intensity of {intensity}/5.
Key symbols included: {symbols}

Style: Dreamlike, surreal, with rich symbolism
"""

# Function to format prompts with dream data
def format_analysis_prompt(dream_data):
    """Format the analysis prompt with dream data."""
    # Create a formatted list of symbols for individual analysis
    symbol_list = "\n".join([f"- {symbol}" for symbol in dream_data.mainSymbols])
    
    return DREAM_ANALYSIS_PROMPT.format(
        narrative=dream_data.narrative,
        symbols=", ".join(dream_data.mainSymbols),
        symbol_list=symbol_list,
        emotion=dream_data.primaryEmotion,
        intensity=dream_data.emotionalIntensity,
        life_connection=dream_data.lifeConnection
    )

def format_image_prompt(dream_data):
    """Format the image generation prompt with dream data."""
    return IMAGE_GENERATION_PROMPT.format(
        image_prompt=dream_data.imagePrompt,
        emotion=dream_data.primaryEmotion,
        intensity=dream_data.emotionalIntensity,
        symbols=", ".join(dream_data.mainSymbols)
    )

File: src/fireworksApi/prompts.js
const DREAM_ANALYSIS_PROMPT = `
As a dream interpreter, analyze the following dream information:

Dream narrative: {narrative}
Main symbols or elements: {symbols}
Primary emotion felt: {emotion}
Emotional intensity (1-5): {intensity}
Connection to waking life: {life_connection}

Provide a thoughtful analysis of what this dream might mean, focusing on emotional insights and 
potential symbolism. Consider the personal context shared by the dreamer and avoid generic 
interpretations. Be insightful yet concise.

Your response must be in JSON format with the following structure:
{
  "interpretation": "Your insightful interpretation of the dream, or an empty string if not determinable",
  "symbols": [
    {
      "title": "Symbol name, or an empty string if not determinable",
      "explanation": "Explanation of what this symbol represents in the dream, or an empty string if not determinable"
    }
  ],
  "advice": "Practical advice for the dreamer based on the interpretation, or an empty string if not determinable",
  "imagePrompt": "Detailed visual description for generating an image, or an empty string if not determinable"
}

Ensure the following:
1. All fields in the JSON structure are present and non-empty (use empty strings or empty arrays if data is not determinable). This is such that zod validation can be performed on the response.
2. The response is valid JSON and does not include any additional text, commentary, or formatting like "\`\`\`JSON".
3. Avoid omitting any fields, even if their values are empty.

Example response:
{
  "interpretation": "This dream suggests...",
  "symbols": [
    {
      "title": "Water",
      "explanation": "Represents emotions and the subconscious"
    }
  ],
  "advice": "Consider exploring your feelings about...",
  "imagePrompt": "A surreal landscape with flowing water, vibrant colors..."
}
`;

const IMAGE_GENERATION_PROMPT = `
Create a dreamlike visualization of the following dream:

{image_prompt}

The dream evoked feelings of {emotion} with an intensity of {intensity}/5.
Key symbols included: {symbols}

Style: Dreamlike, surreal, with rich symbolism
`;

function formatAnalysisPrompt(dreamData) {
  return DREAM_ANALYSIS_PROMPT.replace("{narrative}", dreamData.narrative)
    .replace("{symbols}", dreamData.mainSymbols.join(", "))
    .replace("{emotion}", dreamData.primaryEmotion)
    .replace("{intensity}", dreamData.emotionalIntensity)
    .replace("{life_connection}", dreamData.lifeConnection);
}

function formatImagePrompt(dreamData) {
  return IMAGE_GENERATION_PROMPT.replace(
    "{image_prompt}",
    dreamData.imagePrompt
  )
    .replace("{emotion}", dreamData.primaryEmotion)
    .replace("{intensity}", dreamData.emotionalIntensity)
    .replace("{symbols}", dreamData.mainSymbols.join(", "));
}

module.exports = { formatAnalysisPrompt, formatImagePrompt };
